### PR TITLE
8256655: rework long counted loop handling

### PR DESCRIPTION
**Pull request by rwestrel**
Currently the transformation of a long counted loop into a loop nest
with an inner int counted loop is performed in 2 steps:

1- recognize the counted loop shape and build the loop nest
2- transform the inner loop into a counted loop

I propose changing this to a 3 steps process:

1- recognize the counted loop shape (transform the loop into a LongCountedLoop)
2- build the loop nest
3- transform the inner loop into a counted loop

The benefits are:

- the logic is cleaner because step 1 and 2 are now separated
- Simple optimizations (loop iv type, empty loop elimination, parallel
  iv) can be implemented for LongCountedLoop by refactoring existing
  code

1- above is achieved by refactoring the
PhaseIdealLoop::is_counted_loop() so it takes an extra parameter (the
kind of counted loop, int or long).

2- is the existing loop nest construction logic. But now that it takes
a LongCountedLoop as input, the shape of the loop is known to be that
of a canonicalized counted loop. As a result, the loop nest
construction is simpler.

This change also refactors PhiNode::Value() so that it works for both
CountedLoop and LongCountedLoop.

I also changed ConvL2INode to be a TypeNode (ConvI2LNode is a type
node) and:

    jlong init_p = (jlong)init_t->_lo + stride_con;
    if (init_p > (jlong)max_jint || init_p > (jlong)limit_t->_hi)
      return false; // cyclic loop or this loop trips only once

to:

    if (init_t->lo_as_long() > max_signed_integer(iv_bt) - stride_con) {

because if the loop has a single iteration transforming it to a
CountedLoop should allow the backedge to be optimized out.

/cc hotspot-compiler

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ❌ (2/2 failed) | ❌ (2/2 failed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) |

**Failed test tasks**
- [Linux x64 (build debug)](https://github.com/rwestrel/jdk/runs/1473694989)
- [Linux x64 (build release)](https://github.com/rwestrel/jdk/runs/1473694940)
- [Linux x86 (build debug)](https://github.com/rwestrel/jdk/runs/1473694697)
- [Linux x86 (build release)](https://github.com/rwestrel/jdk/runs/1473694638)

### Issue
 * [JDK-8256655](https://bugs.openjdk.java.net/browse/JDK-8256655): rework long counted loop handling

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to fa2bedf0b2de349a47843b8c9ca83c534dec7e81
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to bd4570a7d19be05471f00dd3c2de64331f1b6069

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1316/head:pull/1316`
`$ git checkout pull/1316`

### CODE REVIEW

File: src/hotspot/share/opto/cfgnode.cpp [hunk 1]
```diff
@@ -1084,30 +1084,32 @@ const Type* PhiNode::Value(PhaseGVN* phase) const {
...
+      const Node* init = l->init_trip();
+      const Node* limit = l->limit();
       const Node* stride = l->stride();
       if (init != NULL && limit != NULL && stride != NULL) {
-        const TypeInt* lo = phase->type(init)->isa_int();
-        const TypeInt* hi = phase->type(limit)->isa_int();
-        const TypeInt* stride_t = phase->type(stride)->isa_int();
+        const TypeInteger* lo = phase->type(init)->isa_integer(l->bt());
+        const TypeInteger* hi = phase->type(limit)->isa_integer(l->bt());
+        const TypeInteger* stride_t = phase->type(stride)->isa_integer(l->bt());
```
Comment (vnkozlov) [inline src/hotspot/share/opto/cfgnode.cpp:1097]: Do we have assert somewhere which checks that all loop's values (init, limit, stride) have the same type?

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
This fails to build on Windows:
`
[2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2065: 'ulong': undeclared identifier
[2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2146: syntax error: missing ')' before identifier 'iters_limit'
[2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2146: syntax error: missing ';' before identifier 'iters_limit'
[2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2065: 'ulong': undeclared identifier
[2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2059: syntax error: ')'`

File: src/hotspot/share/opto/cfgnode.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/cfgnode.cpp:1097]: Thanks for reviewing it.

> Do we have assert somewhere which checks that all loop's values (init, limit, stride) have the same type?

isa_integer() is passed l->bt() which is T_INT for a CountedLoop and T_LONG for LongCountedLoop. This causes isa_integer() to either call isa_int() or isa_long(). So yes, they should all be the same type and the type that matches the loop's type.

**Comment by rwestrel** [comment]
> This fails to build on Windows:
> ` [2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2065: 'ulong': undeclared identifier [2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2146: syntax error: missing ')' before identifier 'iters_limit' [2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2146: syntax error: missing ';' before identifier 'iters_limit' [2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2065: 'ulong': undeclared identifier [2020-11-20T08:11:03,704Z] ./open/src/hotspot/share/opto/loopnode.cpp(821): error C2059: syntax error: ')'`

Build issues should be fixed now.

**Comment by TobiHartmann** [comment]
Did some quick sanity testing and `compiler/regalloc/TestC1OverlappingRegisterHint` fails with `assert(init_n->get_int() + cl->stride_con() >= cl->limit()->get_int()) failed: should be one iteration`.

**Comment by rwestrel** [comment]
> Did some quick sanity testing and `compiler/regalloc/TestC1OverlappingRegisterHint` fails with `assert(init_n->get_int() + cl->stride_con() >= cl->limit()->get_int()) failed: should be one iteration`.

Thanks. That assert is broken AFAICT. It doesn't seem to account for a downward loop (which would seem to indicate that IdealLoopTree::do_one_iteration_loop() never triggered). I pushed a fix.

**Comment by TobiHartmann** [comment]
I ran this through some more stress testing with different values of `StressLongCountedLoop`. It all looks good except for the following crash with `compiler/c2/Test8217359.java`:

`assert(!in->is_CFG()) failed: CFG Node with no controlling input?`
```
Current CompileTask:
C2:    295    2 %           compiler.c2.Test8217359::test @ 43 (164 bytes)

Stack: [0x00007f4abcffc000,0x00007f4abd0fd000],  sp=0x00007f4abd0f7490,  free space=1005k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x1324824]  PhaseIdealLoop::build_loop_early(VectorSet&, Node_List&, Node_Stack&)+0x6e4
V  [libjvm.so+0x132ccca]  PhaseIdealLoop::build_and_optimize(LoopOptsMode)+0x4aa
V  [libjvm.so+0xa16698]  PhaseIdealLoop::optimize(PhaseIterGVN&, LoopOptsMode)+0x328
V  [libjvm.so+0xa12955]  Compile::Optimize()+0x13c5
V  [libjvm.so+0xa147c8]  Compile::Compile(ciEnv*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0x1838
V  [libjvm.so+0x846cec]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1dc
V  [libjvm.so+0xa24a28]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xe08
V  [libjvm.so+0xa25578]  CompileBroker::compiler_thread_loop()+0x5a8
V  [libjvm.so+0x18ab7c6]  JavaThread::thread_main_inner()+0x256
V  [libjvm.so+0x18b2150]  Thread::call_run()+0x100
V  [libjvm.so+0x15950f6]  thread_native_entry(Thread*)+0x116
```

Happens with `-XX:+UseParallelGC -XX:+UseNUMA -XX:StressLongCountedLoop=200000000`. Let me know if you need more info to reproduce.

**Comment by rwestrel** [comment]
> I ran this through some more stress testing with different values of `StressLongCountedLoop`. It all looks good except for the following crash with `compiler/c2/Test8217359.java`:
> 
> `assert(!in->is_CFG()) failed: CFG Node with no controlling input?`
> 
> ```
> Current CompileTask:
> C2:    295    2 %           compiler.c2.Test8217359::test @ 43 (164 bytes)
> 
> Stack: [0x00007f4abcffc000,0x00007f4abd0fd000],  sp=0x00007f4abd0f7490,  free space=1005k
> Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
> V  [libjvm.so+0x1324824]  PhaseIdealLoop::build_loop_early(VectorSet&, Node_List&, Node_Stack&)+0x6e4
> V  [libjvm.so+0x132ccca]  PhaseIdealLoop::build_and_optimize(LoopOptsMode)+0x4aa
> V  [libjvm.so+0xa16698]  PhaseIdealLoop::optimize(PhaseIterGVN&, LoopOptsMode)+0x328
> V  [libjvm.so+0xa12955]  Compile::Optimize()+0x13c5
> V  [libjvm.so+0xa147c8]  Compile::Compile(ciEnv*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0x1838
> V  [libjvm.so+0x846cec]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1dc
> V  [libjvm.so+0xa24a28]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xe08
> V  [libjvm.so+0xa25578]  CompileBroker::compiler_thread_loop()+0x5a8
> V  [libjvm.so+0x18ab7c6]  JavaThread::thread_main_inner()+0x256
> V  [libjvm.so+0x18b2150]  Thread::call_run()+0x100
> V  [libjvm.so+0x15950f6]  thread_native_entry(Thread*)+0x116
> ```
> 
> Happens with `-XX:+UseParallelGC -XX:+UseNUMA -XX:StressLongCountedLoop=200000000`. Let me know if you need more info to reproduce.

Thanks for running tests. I pushed a fix for that one.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. I'll re-run testing and report back once it finished.

File: src/hotspot/share/opto/loopnode.cpp [hunk 2]
```diff
@@ -1240,9 +1238,9 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
     return false;
   }
 
-  if (iv_bt == T_INT &&
-      x->in(LoopNode::LoopBackControl)->Opcode() == Op_SafePoint &&
-      LoopStripMiningIter != 0) {
+  if (x->in(LoopNode::LoopBackControl)->Opcode() == Op_SafePoint &&
+      (iv_bt == T_INT && LoopStripMiningIter != 0) ||
+      iv_bt == T_LONG) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopnode.cpp:1243]: This fails to build on mac:
```
51e416a91494/workspace/open/src/hotspot/share/opto/loopnode.cpp:1241:66: error: '&&' within '||' [-Werror,-Wlogical-op-parentheses]
[2020-11-30T10:06:24,243Z]   if (x->in(LoopNode::LoopBackControl)->Opcode() == Op_SafePoint &&
[2020-11-30T10:06:24,243Z]       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
```

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

File: src/hotspot/share/opto/loopnode.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopnode.cpp]: I pushed a fix for that.

**Review by TobiHartmann — APPROVED** [review]
All tests passed. Looks good to me.

**Comment by rwestrel** [comment]
@TobiHartmann @iwanowww thanks for the reviews

**Pull request closed without merging**